onnxscript/rewriter/pattern.py

```diff
@@ -4,254 +4,156 @@
 import inspect
 import itertools
 import math
-from typing import Any, Callable, List, MutableSequence, Optional, Sequence, Tuple
+from typing import (
+    Any,
+    Callable,
+    List,
+    MutableSequence,
+    Optional,
+    Protocol,
+    Sequence,
+    Tuple,
+    TypeVar,
+    Union,
+)
 
-import numpy as np
 import onnx
-import onnx.numpy_helper
-import onnx.printer
 
 from onnxscript import ir
 from onnxscript.ir import _convenience
 from onnxscript.rewriter import _ir_utils, _tape
 
-# Overview of the pattern module: The classes below are used to define both
-# patterns (that we search for) and replacements for rewrite rules.
-# The matches() method of a pattern is used to check if an IR component
-# matches the pattern.
-# TODO: Ensure that all matches() methods have same type signature (where
-# appropriate).
+T = TypeVar("T")
 
 
-class PythonPattern:
-    def __init__(self, value: int | str | Sequence, name: str | None = None) -> None:
-        self._value = value
-        self._name = name
-
-    @property
-    def value(self) -> int | str | Sequence:
-        return self._value
-
-    @property
-    def name(self) -> str | None:
-        return self._name
-
-    def matches(self, value: int | str | Sequence) -> bool:
-        return value == self.value
-
-
-class StringConstantPattern:
-    def __init__(self, value: str, name: str) -> None:
-        self._value = value
-        self._name = name
-
-    @property
-    def value(self) -> str:
-        return self._value
-
-    @property
-    def name(self) -> str:
-        return self._name
-
-    def matches(self, attr: ir.AttrString) -> bool:
-        return attr.value == self.value
-
-
-class IntConstantPattern:
-    def __init__(self, value: int, name: str) -> None:
-        self._value = value
-        self._name = name
-
-    @property
-    def value(self) -> int:
-        return self._value
+class Pattern(Protocol[T]):  # type: ignore[misc]
+    """This is essentially a Predicate[T], that is, a Callable[[T], bool] bound to the name "matches"."""
 
-    @property
-    def name(self) -> str:
-        return self._name
+    def matches(self, item: T) -> bool: ...
 
-    def matches(self, attr: ir.AttrInt64) -> bool:
-        return attr.value == self.value
 
+class StringConstantPattern(Pattern[str]):
+    """Matches strings with given value."""
```

Comment:
Maybe we should change this name to more domain pattern related?

Reply:
It is just a matching on a string value. It can be used for, eg., both comparing domains, as well as comparing operator names.